### PR TITLE
Update parsers

### PR DESCRIPTION
**Pull request by stec0**
Update of parsers :
- changed bluekai parser rules to handle new data format
- possibility to parse line with custom schemas - new variable in line.parse method
- automatic deployment to pypi by on tagged commit in master with travis - can be done by editing setup.py file and making a github release

Thanks,

### CODE REVIEW

File: artefact/connectors/bluekai/sdt/line.py [hunk 1]
```diff
@@ -125,38 +157,37 @@ def unrolled_lines(cls, payload):
...
-                    'uri_vars': g('uri_vars'),
+                output = {
+                    'bkuuid': g('bkuuid'),
                     'event_date': t[0],
                     'categories': t[1],
                     'event_type': 'event',
-                    'puuid': payload['puuid']
                 }
+                for field in schema:
+                    if field not in ['event_date', 'categories', 'event_type', 'bkuuid']:
```
Comment (lboudard) [inline artefact/connectors/bluekai/sdt/line.py]: 'not in list' not you must :p
https://stackoverflow.com/questions/3949310/how-is-set-implemented

File: artefact/connectors/bluekai/sdt/line.py [hunk 2]
```diff
@@ -125,38 +157,37 @@ def unrolled_lines(cls, payload):
...
-                    'uri_vars': {},
+                output = {
+                    'bkuuid': g('bkuuid'),
                     'event_date': t[0],
                     'categories': t[1],
                     'event_type': 'backlog',
-                    'puuid': payload['puuid']
                 }
+                for field in schema:
+                    if field not in ['event_date', 'categories', 'event_type', 'bkuuid']:
```
Comment (lboudard) [inline artefact/connectors/bluekai/sdt/line.py]: same

File: artefact/connectors/bluekai/sdt/line.py [hunk 3]
```diff
@@ -1,21 +1,62 @@
...
-        (r'^(?P<bkuuid>[^\t|]+)(\t|\|)', False),
-        (r'^(?P<ts_categories>[^\t|]+)(\t|\|)', False),
-        (r'(\t)none|(?P<puuid>[\d]+)$', False),
-        (r'(?P<unused>\|\d+\|(\d+\.){3}\d+)?(\t|\|)?$', False),
-        (r'(?P<referrer>https?:\/\/[^\t|]*)?(\t|\|)?$', False),
-        (r'(?P<unused_url>https?:\/\/[^\t|]*\|?)?(\t|\|)?$', False),
-        (r'^(?P<tag_uri>[^\s\|]*)(\t|\|)?', False)
+        ("^(?P<bkuuid>.{16})(\t|\|| +)", False),
+        ("^(?P<ts_categories>[^\t| ]+)?(\t|\|| +)?", False),
+        ("^(?P<tag_uri>[^\s\|]*)?(\t+|\|| +)?", False),
```
Comment (lboudard) [inline artefact/connectors/bluekai/sdt/line.py]: I wonder, @d-tw , if we go for implicit version detection, sounds like we should add version flag in a headers.
If we go for client calling lib with version, then change to perform at client level as well. WDYT?

File: artefact/connectors/bluekai/sdt/line.py [hunk 4]
```diff
@@ -125,38 +157,37 @@ def unrolled_lines(cls, payload):
 
         sorted_ts_events = sorted(ts_events, key=lambda x: x[0])
 
-        def g(x):
-            return payload.get(x, {})
+        def g(x, default=''):
+            return payload.get(x, default)
```
Comment (lboudard) [inline artefact/connectors/bluekai/sdt/line.py]: I guess this change the default former behavior (that might be without impacts though)

File: artefact/connectors/smart/clicks/line.py [hunk 5]
```diff
@@ -6,41 +6,109 @@
...
+        (";(?P<timestamp_utc>\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2}\\.\\d{3})$", False),
+        (";(?P<filter>[^;]*)$", False),
+        (";(?P<postalcode>[^;]*)$", False),
+        (";(?P<carrier_id>\\d+)$", False),
+        (";(?P<linespeed>[^;]*)$", False),
+        ("^(?P<user_agent>.*)$", False)
     ]
 
-    schema = [
+    default_value = {
```
Comment (lboudard) [inline artefact/connectors/smart/clicks/line.py]: a bunch of repetitions between, default value, schema, rules, required fields, sounds like it should live within yamls
dataset:
    separator: '\t'
    fields:
        toto:
            type: integer
            required: false
            reg: '(\d{6})'

@d-tw  WDYT?

File: artefact/connectors/bluekai/sdt/line.py (hunk 3 above)
Comment (stephanecoquet) [inline artefact/connectors/bluekai/sdt/line.py]: Actually, the parser is implicitly capable of parsing any version of the bluekai data. Consequently, I changed the way a line is parsed a bit - puuid is not being removed, it is being computed later.

**Comment by stephanecoquet** [comment]
@lboudard Some new changes on the parser : 
- Explicit specification of the possible schemas
- Stricter verification of the compatibility of the specified schema with the input data
- Dynamic building of the parsing ruleset

@d-tw Some advances compared to where we left the parser :
- adapted the approach we started to implement
- fixed some bugs (repeat rules, python string formating vs. regex ...)

Thanks!

**Pull request closed without merging**